core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+    primary_key: Option<Vec<usize>>,
+}
+
+impl SizeOf for PersistentState {
+    fn size_of(&self) -> u64 {
+        use std::mem::size_of;
+
+        size_of::<Self>() as u64
+    }
+
+    fn deep_size_of(&self) -> u64 {
+        self.size_of()
+    }
+}
+
+impl State for PersistentState {
+    fn process_records(&mut self, records: &mut Records, partial_tag: Option<Tag>) {
+        assert!(partial_tag.is_none(), "PersistentState can't be partial");
+        if records.len() == 0 {
+            return;
+        }
+
+        let mut batch = WriteBatch::default();
+        for r in records.iter() {
+            match *r {
+                Record::Positive(ref r) => {
+                    self.insert(&mut batch, r.clone());
+                }
+                Record::Negative(ref r) => {
+                    self.remove(&mut batch, r);
+                }
+                Record::BaseOperation(..) => unreachable!(),
+            }
+        }
+
+        // Sync the writes to RocksDB's WAL:
+        let mut opts = rocksdb::WriteOptions::default();
+        opts.set_sync(true);
+        self.db.as_ref().unwrap().write_opt(batch, &opts).unwrap();
+    }
+
+    fn lookup(&self, columns: &[usize], key: &KeyType) -> LookupResult {
+        let key_values = match *key {
+            KeyType::Single(a) => vec![a],
+            KeyType::Double(ref r) => vec![&r.0, &r.1],
+            KeyType::Tri(ref r) => vec![&r.0, &r.1, &r.2],
+            KeyType::Quad(ref r) => vec![&r.0, &r.1, &r.2, &r.3],
+            KeyType::Quin(ref r) => vec![&r.0, &r.1, &r.2, &r.3, &r.4],
+            KeyType::Sex(ref r) => vec![&r.0, &r.1, &r.2, &r.3, &r.4, &r.5],
+        };
+
+        let db = self.db.as_ref().unwrap();
+        let data = match self.primary_key {
+            Some(ref pk) if &pk[..] == columns => {
+                let prefix = Self::serialize_prefix(0, &key_values);
+                // This is a primary key, so we know there's only one row to retrieve
+                // (no need to use prefix_iterator).
+                let raw_row = db.get(&prefix).unwrap();
+                if let Some(raw) = raw_row {
+                    let row: Vec<DataType> = bincode::deserialize(&*raw).unwrap();
+                    vec![Row(Rc::new(row))]
+                } else {
+                    vec![]
+                }
+            }
+            _ => {
+                let index_id = self.get_index_id(columns)
+                    .expect("lookup on non-indexed column set");
+                let prefix = Self::serialize_prefix(index_id, &key_values);
+                let values = db.prefix_iterator(&prefix);
+                if index_id > 0 {
+                    // For non-primary indices we need to first retrieve the primary index key
+                    // through our secondary indices, and then call `.get` again on that.
+                    values
+                        .map(|(_key, value)| {
+                            let raw_row = db.get(&value)
```

Comment:
Is there some kind of `multi_get` we could use here? Might be a lot faster...

Reply:
There is, but it doesn't help with performance:

> Q: If I want to retrieve 10 keys from rocksdb, is it better to batch them and use MultiGet() versus issuing 10 individual Get() calls?
> A: The performance is similar. MultiGet() reads from the same consistent view, but it is not faster.

Might be cleaner regardless though, as it gives us one less FFI call. `rust-rocksdb` doesn't support it yet, but shouldn't be too hard to add - I'll look into it.

Reply:
Ah, I guess I was (erroneously) thinking of RocksDB as being external, and that you'd have to wait for an RTT or something, but that's obviously not true. FFI calls seem fine.